File: 0707-design-linked-list/0707-design-linked-list.py
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class MyLinkedList:

    def __init__(self):
        self.size = 0
        self.head = ListNode(0)

    def get(self, index: int) -> int:
        if index < 0 or index >= self.size:
            return -1
        
        curr = self.head
        for _ in range(index + 1):
            curr = curr.next
        return curr.val

    def addAtHead(self, val: int) -> None:
        self.addAtIndex(0, val)

    def addAtTail(self, val: int) -> None:
        self.addAtIndex(self.size, val)

    def addAtIndex(self, index: int, val: int) -> None:
        if index > self.size:
            return
        
        if index < 0:
            index = 0
        
        self.size += 1
        pred = self.head
        for _ in range(index):
            pred = pred.next
        
        to_add = ListNode(val)
        to_add.next = pred.next
        pred.next = to_add
        

    def deleteAtIndex(self, index: int) -> None:
        if index < 0 or index >= self.size:
            return
        
        self.size -= 1
        pred = self.head
        for _ in range(index):
            pred = pred.next
        pred.next = pred.next.next


# Your MyLinkedList object will be instantiated and called as such:
# obj = MyLinkedList()
# param_1 = obj.get(index)
# obj.addAtHead(val)
# obj.addAtTail(val)
# obj.addAtIndex(index,val)
# obj.deleteAtIndex(index)

File: 0707-design-linked-list/0707-design-linked-list.cpp
class MyLinkedList {
    int sz;
    ListNode *head;
public:
    MyLinkedList() {
        sz = 0;
        head = new ListNode(0);
        
    }
    
    int get(int index) {
        if (index < 0 || index >= sz) return -1;
        ListNode *curr = head;
        
        for (int _ = 0; _ <= index; ++_) curr = curr -> next;
        return curr -> val;
    }
    
    void addAtHead(int val) {
        addAtIndex(0, val);
    }
    
    void addAtTail(int val) {
        addAtIndex(sz, val);
    }
    
    void addAtIndex(int index, int val) {
        if (index > this -> sz) return;
        if (index < 0) index = 0;
        ++sz;
        ListNode* pred = head;
        for (int i = 0; i < index; ++i) pred = pred -> next;
        ListNode* to_add = new ListNode(val);
        to_add -> next = pred -> next;
        pred -> next = to_add;
        
    }
    
    void deleteAtIndex(int index) {
        if (index < 0 || index >= sz) return;
        --sz;
        ListNode *pred = head;
        for (int i = 0; i < index; ++i) pred = pred -> next;
        pred -> next = pred -> next -> next;
    }
};

/**
 * Your MyLinkedList object will be instantiated and called as such:
 * MyLinkedList* obj = new MyLinkedList();
 * int param_1 = obj->get(index);
 * obj->addAtHead(val);
 * obj->addAtTail(val);
 * obj->addAtIndex(index,val);
 * obj->deleteAtIndex(index);
 */